static/src/js/builder.jsx

```diff
@@ -8,13 +8,21 @@ class CodelistBuilder extends React.Component {
   constructor(props) {
     super(props);
 
-    this.state = { updateQueue: [], updating: false };
-
     this.codes = props.hierarchy.nodes;
-    this.codes.forEach((code) => {
-      this.state["status-" + code] = props.codeToStatus[code];
-      this.state["expanded-" + code] = true;
-    });
+
+    // build all code states into a single object
+    const codeStates = Object.fromEntries(
+      this.codes.flatMap((code) => {
+        // build up an array of 2-item arrays to convert to an Object with
+        // fromEntries
+        const status = [`status-${code}`, props.codeToStatus[code]];
+        const expanded = [`expanded-${code}`, true];
+        return [status, expanded];
+      })
+    );
+
+    // Create initial state
+    this.state = { updateQueue: [], updating: false, ...codeStates };
```

Comment:
What was wrong with the old code?

Reply:
You should only set an initial State once per React component.  Previously we had 1 line creating State and two mutating it, eslint rightly called this out.  I tried to convert the lines to `setState()` calls but couldn't get them to work as expected (maybe because they're being called from the constructor??) so opted to switch to one initial creation of state.

Digging into setting state further, [this SO answer](https://stackoverflow.com/a/37760774/158304) (and the `setState()` docs link it contains) lay out the problem quite well I think.  I suspect using `setState()` in the constructor might be related to that method triggering a re-render of the component so it was possibly ignored/batched for later.

Reply:
I understand why fiddling with `state` directly outside of a constructor is a bad thing.  But I don't understand why doing so in the constructor is a problem.

What's the actual eslint error?

I'm all for public immutability, but code is often clearer with some local mutability, including in this case.

Could we do this instead?

```
    let state = { updateQueue: [], updating: false };

    this.codes.forEach((code) => {
      state["status-" + code] = props.codeToStatus[code];
      state["expanded-" + code] = true;
    });

    this.state = state;
```

Reply:
> What's the actual eslint error?

[react/no-direct-mutation-state](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-direct-mutation-state.md)

There's a discussion on the rule [here](https://github.com/yannickcr/eslint-plugin-react/issues/832) which might be of interest, it boils down to "mutation is bad", which I generally agree with (and appears to be the direction the JS community is taking to try and squash the languages various unexpected behaviours).

Given the example above though, that's a very clear block, so lets use that 🙂

Reply:
Fixed in ac6a40a

Reply:
:+1: